ecs-init/backoff/backoff.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2015-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package backoff
+
+import (
+	"math"
+	mathrand "math/rand"
+	"time"
+)
+
+// Backoff defines a back-off/retry interface
+type Backoff interface {
+	Duration() time.Duration
+	ShouldRetry() bool
+}
+
+//go:generate mockgen.sh docker $GOFILE ../docker
+
+type retryBackoff struct {
+	current        time.Duration
+	max            time.Duration
+	jitterMultiple float64
+	multiple       float64
+	maxRetries     int
+	count          int
+}
+
+// NewBackoff creates a Backoff which ranges from min to max increasing by
+// multiple each time.
+// It also adds (and yes, the jitter is always added, never
+// subtracted) a random amount of jitter up to jitterMultiple percent (that is,
+// jitterMultiple = 0.0 is no jitter, 0.15 is 15% added jitter). The total time
+// may exceed "max" when accounting for jitter, such that the absolute max is
+// max + max * jiterMultiple
+func NewBackoff(min, max time.Duration, jitterMultiple, multiple float64, maxRetries int) Backoff {
+	return &retryBackoff{
+		current:        min,
+		max:            max,
+		jitterMultiple: jitterMultiple,
+		multiple:       multiple,
+		maxRetries:     maxRetries,
+		count:          0,
+	}
+}
+
+func (rb *retryBackoff) Duration() time.Duration {
+	ret := rb.current
```

Comment:
I take it that this is not intended to be thread-safe?  Might be worth documenting that a `Backoff` can only be accessed from one goroutine at a time.